chain/network/src/peer_manager/peer_manager_actor.rs

```diff
@@ -1508,62 +1503,38 @@ impl PeerManagerActor {
         if let Some(server_addr) = self.config.addr {
-            // TODO: for now crashes if server didn't start.
-
-            ctx.spawn(TcpListener::bind(server_addr).into_actor(self).then(
-                move |listener, act, ctx| {
-                    let listener = listener.unwrap_or_else(|_| panic!("Failed to PeerManagerActor at {}", server_addr));
-                    let incoming = IncomingCrutch {
-                        listener: tokio_stream::wrappers::TcpListenerStream::new(listener),
-                    };
-                    info!(target: "stats", peer_id = ?act.my_peer_id, addr = ?server_addr, "Server listening at");
-                    let pending_incoming_connections_counter =
-                        act.pending_incoming_connections_counter.clone();
-                    let peer_counter = act.peer_counter.clone();
-                    let max_num_peers: usize = act.config.max_num_peers as usize;
-
-                    ctx.add_message_stream(incoming.filter_map(move |conn| {
-                        if let Ok(conn) = conn {
-                            if pending_incoming_connections_counter.load(Ordering::SeqCst)
-                                + peer_counter.load(Ordering::SeqCst)
-                                < max_num_peers + LIMIT_PENDING_PEERS
-                            {
-                                pending_incoming_connections_counter.fetch_add(1, Ordering::SeqCst);
-                                return future::ready(Some(
-                                    PeerManagerMessageRequest::InboundTcpConnect(
-                                        InboundTcpConnect::new(conn),
-                                    ),
-                                ));
-                            }
-                        }
+            debug!(target: "network", at = ?server_addr, "starting public server");
+            let peer_manager_addr = ctx.address();
+
+            async move {
+                let listener = TcpListener::bind(server_addr).await;
+                let listener = match listener {
+                    Ok(listener) => listener,
+                    Err(e) => {
+                        panic!(
+                            "failed to start listening on server_addr={:?} e={:?}",
+                            server_addr, e
+                        );
+                    }
+                };
 
-                        future::ready(None)
-                    }));
-                    actix::fut::ready(())
-                },
-            ));
+                loop {
+                    if let Ok((conn, client_addr)) = listener.accept().await {
+                        peer_manager_addr.do_send(PeerManagerMessageRequest::InboundTcpConnect(
+                            InboundTcpConnect::new(conn),
+                        ));
+                        debug!(target: "network", from = ?client_addr, "got new connection");
+                    }
+                }
```

Comment:
We don't seem to ever break out of this loop? Ie, when `accept()` returns an error, we will start busy-looping here. It's also not super clear if this handles cancellation for the happy case? How one does actually stop PeerManager?

Reply:
Though it seems like pre-existing logic, so probably is out of scope for the PR.

Reply:
Ok, upon further investigation, it is probably fine -- at the top-level, when we catch `ctlr_c`, we ask the System to stop, and presumably it should stop polling the future.